src/SignalR/clients/csharp/Client.Core/src/HubConnection.cs

```diff
@@ -481,12 +481,31 @@ private async Task StartAsyncCore(CancellationToken cancellationToken)
         var connection = await _connectionFactory.ConnectAsync(_endPoint, cancellationToken).ConfigureAwait(false);
         var startingConnectionState = new ConnectionState(connection, this);
 
+#pragma warning disable CA2252 // This API requires opting into preview features
+        var statefulReconnectFeature = connection.Features.Get<IStatefulReconnectFeature>();
+#pragma warning restore CA2252 // This API requires opting into preview features
+
         // From here on, if an error occurs we need to shut down the connection because
         // we still own it.
         try
         {
-            Log.HubProtocol(_logger, _protocol.Name, _protocol.Version);
-            await HandshakeAsync(startingConnectionState, cancellationToken).ConfigureAwait(false);
+            var usedProtocolVersion = _protocol.Version;
+            if (statefulReconnectFeature is null)
+            {
+                // Stateful Reconnect starts with HubProtocol version 2, newer clients connecting to older servers will fail to connect due to
+                // the handshake only supporting version 1, so we will try to send version 1 during the handshake to keep old servers working.
+                usedProtocolVersion = _protocol.IsVersionSupported(1) ? 1 : _protocol.Version;
+            }
```

Comment:
> And when is _protocol.IsVersionSupported(1) going to return false in practice? It might be better to not even check it.

Probably never, but it does show intent and isn't an expensive call so I'll keep it for now.